mozjs/js/rust/magic_codegen/src/lib.rs

```diff
@@ -244,51 +187,55 @@ struct MagicStructCode {
     js_prop_spec: Vec<quote::Tokens>,
 }
 
-fn gen_structdef_and_slotnum(id: &Option<syn::Ident>,
-                             name: &quote::Ident,
-                             ty: &syn::Ty,
-                             idx: usize) -> (quote::Tokens, quote::Tokens) {
-    let slot_type_name = quote::Ident::from(format!("{}SlotIndex{}", name, idx));
-    let slot = quote! {
-        enum #slot_type_name {}
-        impl SlotIndex for #slot_type_name {
-            fn slot_index() -> u32 { #idx as u32 }
-        }
-    };
-    let field = quote! {
-        #id: MagicSlot<#ty, #slot_type_name>
-    };
-    (field, slot)
-}
-
 fn gen_getter(id: &Option<syn::Ident>,
               name: &quote::Ident,
-              ty: &syn::Ty) -> (quote::Tokens, quote::Ident) {
+              idx: u32) -> (quote::Tokens, quote::Ident) {
     let getter_str = match *id {
         Some(ref real_id) => format!("get_{}", real_id.to_string()),
         None => panic!("Encounter a empty field. Something wrong..."),
     };
     let getter_name = quote::Ident::from(getter_str);
     let getter = quote! {
-        pub unsafe fn #getter_name (obj: &#name, cx: *mut jsapi::JSContext) -> #ty {
-            (*obj).#id.get(cx)
+        pub unsafe fn #getter_name<T: ToJSValConvertible + FromJSValConvertible<Config = ()>> (
+            obj: &#name, cx: *mut jsapi::JSContext) -> T {
+            let jsobj = obj.object;
+            rooted!(in(cx) let val = jsapi::JS_GetReservedSlot(jsobj, #idx));
+
+            let conversion = FromJSValConvertible::from_jsval(cx, val.handle(), ())
+                .expect("Should never put anything into a MagicSlot that we can't \
+                         convert back out again");
+
+            match conversion {
+                ConversionResult::Success(v) => v,
+                ConversionResult::Failure(why) => {
+                    panic!("Should never put anything into a MagicSlot that we \
+                            can't convert back out again: {}",
+                           why);
+                }
+            }
```

Comment:
A couple things here:

* We can't be returning *any* `T: ToJSValConvertible + FromJSValConvertible<Config = ()>` -- it had better be the type we put into the slot! (what used to be `#ty`)

* The getters and setters should be methods of `#name`, not take a reference to a `#name`. This could be a follow up PR.